aws/resource_aws_servicecatalog_product.go

```diff
@@ -0,0 +1,447 @@
+
+	input := &servicecatalog.CreateProductInput{
+		IdempotencyToken: aws.String(resource.UniqueId()),
+		Name:             aws.String(d.Get("name").(string)),
+		Owner:            aws.String(d.Get("owner").(string)),
+		ProductType:      aws.String(d.Get("type").(string)),
+		ProvisioningArtifactParameters: expandProvisioningArtifactParameters(
+			d.Get("provisioning_artifact_parameters").([]interface{})[0].(map[string]interface{}),
+		),
+	}
+
+	if v, ok := d.GetOk("accept_language"); ok {
+		input.AcceptLanguage = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("distributor"); ok {
+		input.Distributor = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("support_description"); ok {
+		input.SupportDescription = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("support_email"); ok {
+		input.SupportEmail = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("support_url"); ok {
+		input.SupportUrl = aws.String(v.(string))
+	}
+
+	if len(tags) > 0 {
+		input.Tags = tags.IgnoreAws().ServicecatalogTags()
+	}
+
+	var output *servicecatalog.CreateProductOutput
+	err := resource.Retry(iamwaiter.PropagationTimeout, func() *resource.RetryError {
+		var err error
+
+		output, err = conn.CreateProduct(input)
+
+		if tfawserr.ErrMessageContains(err, servicecatalog.ErrCodeInvalidParametersException, "profile does not exist") {
+			return resource.RetryableError(err)
+		}
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		output, err = conn.CreateProduct(input)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error creating Service Catalog Product: %w", err)
+	}
+
+	if output == nil {
+		return fmt.Errorf("error creating Service Catalog Product: empty response")
+	}
+
+	if output.ProductViewDetail == nil || output.ProductViewDetail.ProductViewSummary == nil {
+		return fmt.Errorf("error creating Service Catalog Product: no product view detail or summary")
+	}
+
+	if output.ProvisioningArtifactDetail == nil {
+		return fmt.Errorf("error creating Service Catalog Product: no provisioning artifact detail")
+	}
+
+	d.SetId(aws.StringValue(output.ProductViewDetail.ProductViewSummary.ProductId))
+
+	if _, err := waiter.ProductReady(conn, aws.StringValue(input.AcceptLanguage),
+		aws.StringValue(output.ProductViewDetail.ProductViewSummary.ProductId)); err != nil {
+		return fmt.Errorf("error waiting for product to be ready: %w", err)
+	}
+
+	return resourceAwsServiceCatalogProductRead(d, meta)
+}
+
+func resourceAwsServiceCatalogProductRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	output, err := waiter.ProductReady(conn, d.Get("accept_language").(string), d.Id())
```

Comment:
Why is `waiter.ProductReady` called here rather than the underlying `finder`-type function?

Reply:
I guess 2 reasons. 1) There isn't a finder. 2) With the status/waiter/finder pattern still evolving, I'm the least sold on the merits of finders. I'm not necessarily against them but not sure they add that much until we figure out exactly how they contribute in the overall pattern. Particularly, errors need to be handled differently (mostly not found) in different contexts. The finder I like best (because it's easy to work with) so far is the one that just finds and returns errors without editorializing (changing not found to `""` or a new error). Adding layers of complexity isn't always bad but needs a compelling reason that I've not been convinced of yet.